extensions/amp-lightbox/1.0/lightbox.js

```diff
@@ -0,0 +1,212 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {ContainWrapper} from '../../../src/preact/component';
+import {forwardRef} from '../../../src/preact/compat';
+import {setStyle} from '../../../src/style';
+import {
+  useEffect,
+  useImperativeHandle,
+  useRef,
+  useState,
+} from '../../../src/preact';
+import {useStyles} from './lightbox.jss';
+
+const ANIMATION_DURATION = 200;
+const ANIMATION_PRESETS = {
+  'fade-in': [
+    {opacity: 0, visibility: 'visible'},
+    {opacity: 1, visibility: 'visible'},
+  ],
+  'fly-in-top': [
+    {opacity: 0, transform: 'translate(0,-100%)', visibility: 'visible'},
+    {opacity: 1, transform: 'translate(0, 0)', visibility: 'visible'},
+  ],
+  'fly-in-bottom': [
+    {opacity: 0, transform: 'translate(0, 100%)', visibility: 'visible'},
+    {opacity: 1, transform: 'translate(0, 0)', visibility: 'visible'},
+  ],
+};
+
+// Note: Sometimes, the timing between render and lightboxRef update is delayed, so we have a quick function to wait for the ref element before acting on it.
+/**
+ * @param {Object} ref
+ * @param {number} time
+ * @return {number}
+ */
+function getRefElement(ref, time = 1) {
+  return new Promise((resolve, reject) => {
+    if (time <= 0) {
+      return reject(`${ref} not found`);
+    }
+    if (ref) {
+      return resolve(ref);
+    }
+    setTimeout(() => getRefElement(ref, time - 1).then(resolve), 100);
+  });
+}
+
+/**
+ * @param {!LightboxDef.Props} props
+ * @param {{current: (!LightboxDef.LightboxApi|null)}} ref
+ * @return {PreactDef.Renderable}
+ */
+function LightboxWithRef(
+  {
+    id,
+    animateIn = 'fade-in',
+    closeButtonAriaLabel,
+    children,
+    initialOpen,
+    beforeOpen,
+    afterClose,
+    ...rest
+  },
+  ref
+) {
+  // There are two phases to open and close.
+  // To open, we mount and render the contents (invisible), then animate the display (visible).
+  // To close, it's the reverse.
+  // `mounted` mounts the component. `visible` plays the animation.
+  const [mounted, setMounted] = useState(initialOpen);
+  const [visible, setVisible] = useState(false);
+  const classes = useStyles();
+  const lightboxRef = useRef();
+
+  // We are using refs here to refer to common strings, objects, and functions used.
+  // This is because they are needed within `useEffect` calls below (but are not depended for triggering)
+  // This bypasses the exhaustive-deps eslint error
+  const animateInRef = useRef();
+  animateInRef.current = animateIn;
+
+  useImperativeHandle(
+    ref,
+    () => ({
+      open: () => {
+        if (beforeOpen) {
+          beforeOpen();
+        }
+        setMounted(true);
+        setVisible(true);
+      },
+      close: () => {
+        setVisible(false);
+      },
+    }),
+    [beforeOpen]
```

Comment:
Also not great to recreate this API each time this handler changes. Would also be better to pass it via ref. Let's create a `useValueRef` hook in this module like this:

```
function useValueRef(value) {
  const ref = useRef();
  ref.current = value;
  return ref;
}
```

Then, our code will look a bit simpler:

```
const animateInRef = useValueRef(animateIn);
const onBeforeOpenRef = useValueRef(onBeforeOpen);
const onAfterCloseRef = useValueRef(onAfterClose);
```